irods-vfs-impl/src/main/java/org/irods/nfsrods/vfs/IRODSVirtualFileSystem.java

```diff
@@ -1645,15 +1666,118 @@ private static Inode toFh(long _inodeNumber)
+                }                                    
+            } 
+            catch (GenQueryBuilderException | JargonException | JargonQueryException e)
+            {
+                log_.error(e.getMessage());
+            }    
+        } 
+        else if (ObjectType.DATA_OBJECT == _option) 
+        {
+            try 
+            {
+                builder = new IRODSGenQueryBuilder(true, null);
+    
+                builder.addSelectAsGenQueryValue(RodsGenQueryEnum.COL_D_DATA_ID)
+                    .addSelectAsGenQueryValue(RodsGenQueryEnum.COL_COLL_NAME)
+                    .addSelectAsGenQueryValue(RodsGenQueryEnum.COL_DATA_NAME)
+                    .addConditionAsGenQueryField(RodsGenQueryEnum.COL_D_DATA_ID, QueryConditionOperators.EQUAL, String.valueOf(_inodeNumber));
+                
+                // the constructed GenQuery is: SELECT DATA_ID, COLL_NAME, DATA_NAME WHERE DATA_ID = '<_inodeNumber>'
+    
+                gqe = factory_.getIRODSGenQueryExecutor(adminAcct_);
+                query = builder.exportIRODSQueryFromBuilder(50);
+                resultSet = gqe.executeIRODSQueryAndCloseResultInZone(query, 0, adminAcct_.getZone());
+                results = resultSet.getResults();
+                log_.debug("findLogicalPathByInodeNumber - Found row is (data id) [{}]", results);
+                if (!results.isEmpty()) 
+                {
+                    // since DATA_NAME is only the filename (no path), we'll need to combine it with COLL_NAME to get the full path
+                    // column 1 = COLL_NAME, column 2 = DATA_NAME
+                    strPath = results.get(0).getColumn(1) + '/' + results.get(0).getColumn(2);   
+                    
+                    log_.debug("findLogicalPathByInodeNumber - Combined path is [{}]", strPath);
+                }                       
+            } 
+            catch (GenQueryBuilderException | JargonException | JargonQueryException e)
+            {
+                log_.error(e.getMessage());
+            }
+        } 
+        else 
+        {
+            log_.warn("findLogicalPathByInodeNumber - Option [{}] is not supported. choose from `ObjectType.COLLECTION` or `ObjectType.DATA_OBJECT`", _option);
+        }
+
+        return Optional.ofNullable(strPath);
+    }
+
     private Path getPath(long _inodeNumber) throws IOException
     {
         Path path = inodeToPathMapper_.getPathByInodeNumber(_inodeNumber);
 
+        // If no path matches _inodeNumber, we'll need to find where we are from GenQuery.
+        // Since a path may lead to a data object or a collection, we'll need to check both when using GenQuery.
         if (path == null)
         {
-            throw new NoEntException("Path does not exist for [" + _inodeNumber + "]");
-        }
+            log_.debug("getPath - Path was null (maybe because the NFSRODS container was restarted without remounting), looking up in GenQuery");
```

Comment:
they may not run things in a container...

how about...

"because the NFSRODS container" -> "because NFSRODS"